### PR TITLE
JAMES-3057 mailbox mapper rename refactoring

### PR DESCRIPTION
**Pull request by Arsnael**
Following the work on https://github.com/linagora/james-project/pull/3102

### CODE REVIEW

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMailboxMapper.java [hunk 1]
```diff
@@ -159,6 +160,42 @@ public boolean hasChildren(Mailbox mailbox, char delimiter) throws MailboxExcept
...
+    public MailboxId create(Mailbox mailbox) throws MailboxException {
+        Preconditions.checkArgument(mailbox.getMailboxId() == null, "A mailbox we want to create should not have a mailboxId set already");
+
+        MaildirId maildirId = MaildirId.random();
+        MaildirFolder folder = maildirStore.createMaildirFolder(mailbox);
+
+        if (!folder.exists()) {
+            boolean success = folder.getRootFile().exists();
+            if (!success) {
+                success = folder.getRootFile().mkdirs();
```
Comment (Arsnael) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMailboxMapper.java]: please don't reuse success variable, it makes the code too complicated (from @mbaechler )

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 2]
```diff
@@ -82,6 +83,26 @@ protected void commit() throws MailboxException {
             throw new MailboxException("Commit of transaction failed", e);
         }
     }
+
+    @Override
+    public MailboxId create(Mailbox mailbox) throws MailboxException {
+        Preconditions.checkArgument(mailbox.getMailboxId() == null, "A mailbox we want to create should not have a mailboxId set already");
+
+        try {
+            if (isPathAlreadyUsedByAnotherMailbox(mailbox)) {
```
Comment (Arsnael) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: do we have a transaction opened?
Can't we rely on the database to detect duplicate insertion instead?
(from @mbaechler )

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperConcurrencyTest.java [hunk 3]
```diff
@@ -64,7 +64,7 @@ void setUp(CassandraCluster cassandra) {
     @Test
     void saveShouldBeThreadSafe() throws Exception {
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperConcurrencyTest.java]: rename

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java [hunk 4]
```diff
@@ -90,12 +90,12 @@ void setUp() {
     @Disabled("JAMES-2514 Cassandra 3 supports long mailbox names. Hence we can not rely on this for failing")
     @Test
     void saveShouldNotRemoveOldMailboxPathWhenCreatingTheNewMailboxPathFails() throws Exception {
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java]: idem

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/MailboxPathV2MigrationTest.java [hunk 5]
```diff
@@ -93,15 +93,15 @@ void setUp(CassandraCluster cassandra) {
...
+        mailboxMapper.rename(MAILBOX_1);
 
         assertThat(daoV2.retrieveId(MAILBOX_PATH_1).blockOptional())
             .contains(new CassandraIdAndPath(MAILBOX_ID_1, MAILBOX_PATH_1));
     }
 
     @Test
     void newValuesShouldNotBeSavedInOldDAO() throws Exception {
-        mailboxMapper.save(MAILBOX_1);
+        mailboxMapper.rename(MAILBOX_1);
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/MailboxPathV2MigrationTest.java]: Isn't it a create?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/MailboxPathV2MigrationTest.java [hunk 6]
```diff
@@ -93,15 +93,15 @@ void setUp(CassandraCluster cassandra) {
 
     @Test
     void newValuesShouldBeSavedInMostRecentDAO() throws Exception {
-        mailboxMapper.save(MAILBOX_1);
+        mailboxMapper.rename(MAILBOX_1);
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/MailboxPathV2MigrationTest.java]: Isn't it a create?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java [hunk 7]
```diff
@@ -90,12 +90,12 @@ void setUp() {
...
-        testee.save(new Mailbox(MAILBOX_PATH, UID_VALIDITY));
+        testee.rename(new Mailbox(MAILBOX_PATH, UID_VALIDITY));
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperTest.java]: Isn't it a create?

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 8]
```diff
@@ -178,7 +178,7 @@ void setup() throws MailboxException, IOException {
         session = sessionProvider.createSystemSession(USERNAME);
 
         mailbox = new Mailbox(MailboxPath.forUser(USERNAME, DefaultMailboxes.INBOX), MAILBOX_ID.id);
-        mapperFactory.getMailboxMapper(session).save(mailbox);
+        mapperFactory.getMailboxMapper(session).rename(mailbox);
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: Isn't it a create?

File: mailbox/plugin/spamassassin/src/test/java/org/apache/james/mailbox/spamassassin/SpamAssassinListenerTest.java [hunk 9]
```diff
@@ -88,12 +88,12 @@ void setup() throws Exception {
...
-        mailboxId2 = mailboxMapper.save(mailbox2);
-        spamMailboxId = mailboxMapper.save(new Mailbox(MailboxPath.forUser(USER, "Spam"), UID_VALIDITY));
-        spamCapitalMailboxId = mailboxMapper.save(new Mailbox(MailboxPath.forUser(USER, "SPAM"), UID_VALIDITY));
-        trashMailboxId = mailboxMapper.save(new Mailbox(MailboxPath.forUser(USER, "Trash"), UID_VALIDITY));
+        mailboxMapper.rename(inbox);
+        mailboxId1 = mailboxMapper.rename(mailbox1);
+        mailboxId2 = mailboxMapper.rename(mailbox2);
+        spamMailboxId = mailboxMapper.rename(new Mailbox(MailboxPath.forUser(USER, "Spam"), UID_VALIDITY));
+        spamCapitalMailboxId = mailboxMapper.rename(new Mailbox(MailboxPath.forUser(USER, "SPAM"), UID_VALIDITY));
+        trashMailboxId = mailboxMapper.rename(new Mailbox(MailboxPath.forUser(USER, "Trash"), UID_VALIDITY));
```
Comment (chibenwa) [inline mailbox/plugin/spamassassin/src/test/java/org/apache/james/mailbox/spamassassin/SpamAssassinListenerTest.java]: Isn't it a create?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java [hunk 10]
```diff
@@ -38,11 +38,16 @@
  *
  */
 public interface MailboxMapper extends Mapper {
-    
+
+    /**
+     * Create the given {@link Mailbox} to the underlying storage
+     */
+    MailboxId create(Mailbox mailbox) throws MailboxException;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java]: I bet a next step will be `Mailbox create(MailboxPath, UidValidity)`

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MailboxMapperACLTest.java [hunk 11]
```diff
@@ -58,7 +58,7 @@ void setUp() throws Exception {
         mailboxMapper = createMailboxMapper();
         MailboxPath benwaInboxPath = MailboxPath.forUser(Username.of("benwa"), "INBOX");
         benwaInboxMailbox = createMailbox(benwaInboxPath);
-        mailboxMapper.save(benwaInboxMailbox);
+        mailboxMapper.rename(benwaInboxMailbox);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MailboxMapperACLTest.java]: create

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 12]
```diff
@@ -960,7 +960,7 @@ void deletesShouldNotFailUponMissingMessage() {
     private Mailbox createMailbox(MailboxPath mailboxPath) throws MailboxException {
         Mailbox mailbox = new Mailbox(mailboxPath, UID_VALIDITY);
         mailbox.setMailboxId(mapperProvider.generateId());
-        mailboxMapper.save(mailbox);
+        mailboxMapper.rename(mailbox);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Isn't it a create?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java [hunk 13]
```diff
@@ -1190,7 +1190,7 @@ private Mailbox createMailbox(MailboxPath mailboxPath) throws MailboxException {
         Mailbox mailbox = new Mailbox(mailboxPath, UID_VALIDITY);
         mailbox.setMailboxId(mapperProvider.generateId());
         
-        mailboxMapper.save(mailbox);
+        mailboxMapper.rename(mailbox);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: Isn't it a create?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMoveTest.java [hunk 14]
```diff
@@ -138,7 +138,7 @@ private Mailbox createMailbox(MailboxPath mailboxPath) throws MailboxException {
         Mailbox mailbox = new Mailbox(mailboxPath, UID_VALIDITY);
         MailboxId id = mapperProvider.generateId();
         mailbox.setMailboxId(id);
-        mailboxMapper.save(mailbox);
+        mailboxMapper.rename(mailbox);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMoveTest.java]: Isn't it a create?

File: server/container/mailbox-jmx/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java [hunk 15]
```diff
@@ -65,44 +65,44 @@ void setUp() throws Exception {
 
     @Test
     void deleteMailboxesShouldDeleteMailboxes() throws Exception {
-        mapperFactory.createMailboxMapper(session).save(new Mailbox(MailboxPath.forUser(USER, "name"), UID_VALIDITY));
+        mapperFactory.createMailboxMapper(session).rename(new Mailbox(MailboxPath.forUser(USER, "name"), UID_VALIDITY));
```
Comment (chibenwa) [inline server/container/mailbox-jmx/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: Isn't it a create? (& everywhere else in this file)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MailboxMapperTest.java [hunk 16]
```diff
@@ -110,35 +108,54 @@ void createShouldThrowWhenMailboxAlreadyExists() throws MailboxException {
...
         MailboxId mailboxId = mailboxMapper.create(benwaInboxMailbox);
 
         assertThat(mailboxId).isNotNull();
     }
 
     @Test
     void createShouldThrowWhenMailboxIdNotNull() {
+        benwaInboxMailbox.setMailboxId(generateId());
+
         assertThatThrownBy(() -> mailboxMapper.create(benwaInboxMailbox))
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MailboxMapperTest.java]: I think here the test subject is more "trying to rename a non existing mailbox"

Because by definition you CAN create a non existing mailbox :trollface:

Comment (Arsnael) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MailboxMapperTest.java]: It's a test on checking that create throws if a mailboxId is set up... nothing with rename?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java (hunk 10 above)
Comment (Arsnael) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java]: Other PR :)

**Comment by trantienduchn** [comment]
With the current implementation, If I rename a not-created mailbox, what will happen?

**Comment by chibenwa** [comment]
`If I rename a not-created mailbox, what will happen?`

It should throw.

Do we have a test for this @Arsnael ?

**Comment by Arsnael** [comment]
> `If I rename a not-created mailbox, what will happen?`
> 
> It should throw.
> 
> Do we have a test for this @Arsnael ?

Yes it should throw, as the work I'm doing here is completely separate the logic of create and rename mailbox in all implementations. So after that you shouldn't be able to 'rename' a non existing mailbox. 

Yes there are tests for that coming up in the WIP. And I'm trying to refactor the code around to apply this logic and have it green :)

**Comment by Arsnael** [comment]
I think I more or less finished. Let's see if the CI complains about somewhere I forgot or not :)

**Comment by Arsnael** [comment]
```
[48098addac3b3e1af179e1f7606546c679c64365] [ERROR]   JpaMessageMoveTest>MessageMoveTest.setUp:73->MessageMoveTest.createMailbox:138 » InvalidState
[48098addac3b3e1af179e1f7606546c679c64365] [ERROR]   JpaMessageMoveTest>MessageMoveTest.setUp:73->MessageMoveTest.createMailbox:138 » InvalidState
[48098addac3b3e1af179e1f7606546c679c64365] [ERROR]   JpaMessageMoveTest>MessageMoveTest.setUp:73->MessageMoveTest.createMailbox:138 » InvalidState
[48098addac3b3e1af179e1f7606546c679c64365] [ERROR]   JpaMessageMoveTest>MessageMoveTest.setUp:73->MessageMoveTest.createMailbox:138 » InvalidState
[48098addac3b3e1af179e1f7606546c679c64365] [ERROR]   JpaMessageMoveTest>MessageMoveTest.setUp:73->MessageMoveTest.createMailbox:138 » InvalidState
```

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 17]
```diff
@@ -88,9 +88,7 @@ public MailboxId create(Mailbox mailbox) throws MailboxException {
         Preconditions.checkArgument(mailbox.getMailboxId() == null, "A mailbox we want to create should not have a mailboxId set already");
 
         try {
-            begin();
```
Comment (mbaechler) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: why did you remove all the transaction related calls?

Comment (Arsnael) [reply mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: Because it was crashing a lot of tests in mailbox-jpa module? Getting a lot of those exceptions :

```
org.apache.openjpa.persistence.InvalidStateException: This operation cannot be performed while a Transaction is active.
```

Which I believe is related (even with putting a rollback on a finally block)

So I'm not familiar with the JPA part of the code but what I noticed : 
* in save method before (that was doing create and rename), you didn't have those transaction calls... must have had a reason right?
* `JPAMailboxMapper` extends `JPATransactionalMapper` that I believe defines those transaction calls
* the `execute` of the `TransactionalMapper` (that `JPATransactionalMapper` extends) is executing a transaction with those transactional calls (begin, then running the transaction, then commit, and rollback if it fails). Those are being used in the upper levels it seems (`MailboxManager` level) and the methods called by the mapper are part of the transactions defined in execute, for example create a mailbox on the manager level:

```
mapper.execute(Mapper.toTransaction(() -> mailboxIds.add(mapper.create(mailbox))));
```

Which is maybe why I get those ugly `InvalidStateException` when I add transaction calls inside the jpa mapper methods?

WDYT?

**Comment by Arsnael** [comment]
```
[56a9c40d66597008cbb3fc67cf51544eacf48c5b] [ERROR]   FullUserMaildirMailboxManagerTest$MailboxNameLimitTests>MailboxManagerTest$MailboxNameLimitTests.renamingMailboxShouldNotThrowWhenNameWithoutEmptyHierarchicalLevel:394 
```

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java (hunk 17 above)
Comment (chibenwa) [reply mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: Yes IMO the way JPA manages transactions might be suspicious but it is orthogonal to your work, as such I agree keeping things unchanged here.

File: mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/FullUserMaildirMailboxManagerTest.java [hunk 18]
```diff
@@ -35,52 +34,6 @@
...
-
-        @Disabled("JAMES-2993 mailboxId support for Maildir is partial")
-        @Test
-        protected void getMailboxByIdShouldReturnMailboxWhenBelongingToUser() {
-        }
-    }
-
-    @Nested
-    class MailboxNameLimitTests extends MailboxManagerTest<StoreMailboxManager>.MailboxNameLimitTests {
-        @Disabled("MAILBOX-389 Mailbox rename fails with Maildir")
```
Comment (chibenwa) [inline mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/FullUserMaildirMailboxManagerTest.java]: Cool! Don't forget to notify that you have been solving these maildir issues in the related ticket.

**Comment by Arsnael** [comment]
```
[6c3cac0d8ad605156f06aa50a3972f4eae10cbf7] [ERROR] Errors: 
[6c3cac0d8ad605156f06aa50a3972f4eae10cbf7] [ERROR]   MemorySpamAssassinContractTest.expungingSpamMessageShouldNotImpactSpamAssassinState » ConditionTimeout
```

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
```
[6c3cac0d8ad605156f06aa50a3972f4eae10cbf7] [ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.5.1:buildTar (docker packaging) on project mock-smtp-server: com.google.cloud.tools.jib.api.InsecureRegistryException: Failed to verify the server at https://registry-1.docker.io/v2/library/openjdk/blobs/sha256:9a0b0ce99936ce4861d44ce1f193e881e5b40b5bf1847627061205b092fa7f1d because only secure connections are allowed. -> [Help 1]
```
-_-

test this please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 19]
```diff
@@ -172,49 +172,48 @@ public MailboxId create(Mailbox mailbox) throws MailboxException {
 
         CassandraId cassandraId = CassandraId.timeBased();
         mailbox.setMailboxId(cassandraId);
-        if (!tryCreate(mailbox, cassandraId).block()) {
+        if (!tryCreate(mailbox, cassandraId)) {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: I prefer enforcing block in the final method.

This makes `a single mapper call SHOULD not call several time block` review easier.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 20]
```diff
@@ -172,49 +172,48 @@ public MailboxId create(Mailbox mailbox) throws MailboxException {
...
+        try {
+            return mailboxDAO.retrieveMailbox(cassandraId)
+                .flatMap(mailbox -> mailboxPathV2DAO.save(cassandraMailbox.generateAssociatedPath(), cassandraId)
+                    .filter(isCreated -> isCreated)
+                    .flatMap(mailboxHasCreated -> mailboxPathV2DAO.delete(mailbox.generateAssociatedPath())
+                        .then(mailboxDAO.save(cassandraMailbox))
+                        .thenReturn(true))
+                    .switchIfEmpty(Mono.just(false)))
+                .switchIfEmpty(Mono.error(() -> new MailboxNotFoundException(cassandraId)))
+                .block();
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: Idem, I would prefer us to explicitly block in `tryRename`

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java (hunk 10 above)
Comment (Arsnael) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java]: #3128

**Comment by Arsnael** [comment]
Merged
